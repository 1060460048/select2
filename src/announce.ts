export function initialize() {
    if (document.getElementById("s2-live")) {
        return;
    }

    const live = document.createElement("div");
    live.setAttribute("id", "s2-live");
    live.setAttribute("class", "s2-offscreen s3-live");
    document.body.appendChild(live);

    const assertive = document.createElement("div");
    assertive.setAttribute("id", "s2-live-assertive");
    assertive.setAttribute("role", "log");
    assertive.setAttribute("aria-live", "assertive");
    assertive.setAttribute("aria-relevant", "additions");
    live.appendChild(assertive);

    const polite = document.createElement("div");
    polite.setAttribute("id", "s2-live-polite");
    polite.setAttribute("role", "log");
    polite.setAttribute("aria-live", "polite");
    polite.setAttribute("aria-relevant", "additions");
    live.appendChild(polite);

}

export function assertively(message: string) {
    const node = document.createElement("div");
    node.appendChild(document.createTextNode(message));
    document.getElementById("s2-live-assertive").appendChild(node);
    // TODO clean up old nodes


    // for debugging
    const live = document.getElementById("s2-live");
    live.scrollTop = live.scrollHeight;

}

export function politely(message: string) {
    const node = document.createElement("div");
    node.appendChild(document.createTextNode(message));
    document.getElementById("s2-live-polite").appendChild(node);


    // for debugging
    const live = document.getElementById("s2-live");
    live.scrollTop = live.scrollHeight;

    // TODO clean up old nodes
}
