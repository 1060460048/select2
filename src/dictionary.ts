export interface Dictionary {
    valueAdded: (itemLabel: string) => string;

    noSearchResults: () => string;
    searchResultsLoading: () => string;
    removeButtonTitle: () => string;
    minimumCharactersMessage: (len: number, min: number) => string;
}

class AmericanEnglishDictionary implements Dictionary {

    noSearchResults() {
        return "No results available";
    }

    searchResultsLoading() {
        return "Loading...";
    }

    removeButtonTitle() {
        return "Remove selected values";
    }

    valueAdded(itemLabel: string) {
        return itemLabel + " added";
    }

    minimumCharactersMessage(len: number, min: number) {
        const delta = min - len;
        return "Please enter " + delta + " more character" + (delta > 1 ? "s" : "");
    }
}

const dictionaries = new Map<string, Dictionary>();
dictionaries.set("en_us", new AmericanEnglishDictionary());

export function dictionary(dict: Dictionary | string): Dictionary {
    const fallback = dictionaries.get("en_us");

    if (!dict) {
        return fallback;
    }

    if (typeof dict == "string") {
        const instance = dictionaries.get(dict);
        return instance ? instance : fallback;
    } else {
        return dict;
    }
}
