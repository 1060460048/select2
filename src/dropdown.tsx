import * as React from "react";
import * as util from "./util";
import {createPortal} from "react-dom";

export interface DropdownProps {
    control: HTMLElement, // TODO make this a ref
    className?: string,
    children: React.ReactNode
}

export class Dropdown extends React.PureComponent<DropdownProps, {}> {

    container: HTMLElement;

    scrollParents: EventTarget[];
    private readonly throttledPosition: () => void;

    constructor(props: DropdownProps) {
        super(props);
        this.throttledPosition = util.throttle(50, this.position);
    }

    componentWillMount() {
        this.container = document.createElement("div");
        const cn = this.props.className;
        if (cn) {
            this.container.setAttribute("class", cn);
        }
        document.body.appendChild(this.container);

    }

    componentWillUnmount() {
        this.scrollParents.forEach((parent) => {
            ["resize", "scroll", "touchmove"].forEach((event) => {
                parent.removeEventListener(event, this.throttledPosition);
            });
        });
        delete this.scrollParents;
        document.body.removeChild(this.container);
        delete this.container;
    }


    render() {
        return createPortal(this.props.children, this.container);
    }

    componentDidMount() {
        this.scrollParents = util.getScrollParents(this.props.control);
        this.scrollParents.forEach((parent) => {
            ["resize", "scroll", "touchmove"].forEach((event) => {
                parent.addEventListener(event, this.throttledPosition);
            });
        });
        this.position();
    }

    componentDidUpdate() {
        this.position();
    }

    position = () => {
        const control = this.props.control;
        const rect = control.getBoundingClientRect();
        const style = `
            top: ${rect.top + rect.height + window.pageYOffset}px;
            left: ${rect.left + window.pageXOffset}px;
            width: ${rect.width}px;
        `;
        // TODO also called on scroll of children - should not be
        // console.log("position");
        this.container.setAttribute("style", style);
    }
}

