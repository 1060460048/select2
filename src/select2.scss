
$color-focus: #4D90FE;
$color-item-bkg: #0073E6;
$color-item-text: #fff;
$border: 1px solid #ddd;
$border-focus: 1px solid $color-focus;
$border-radius: 5px;
$dropdown-padding: 8px 10px;

/*
 content: "✓";
content: "×";
*/

.s25-control {
  position: relative;
  box-sizing: border-box;
  min-height: 36px;
  width: 100%;

}

.s25-multi {
  display: flex;
  flex-wrap: nowrap;
}

.s25-multi .s25-body {
  flex-grow: 1;
  display: flex;
  flex-wrap: wrap;
}

.s25-multi .s25-multi-values {
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
  
	.s25-item {
		cursor: pointer;
		box-sizing: border-box;
		align-items: center;
	}
}

.s25-multi div {
  max-width: 100%;
}

.s25-multi .s25-search {
  max-width: 100%;
  min-width: 75px;
  border: none;
}

.s25-multi .s25-toggle {
  display: flex;
  box-sizing: border-box;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 30px;
}

.s25-multi .s25-remove {
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  
	&:disabled {
		cursor: default;
	}
	
	&:active {
		border: 0px;
	}
  
}

.s25-offscreen {
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  width: 1px;
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
}

.s25-dropdown {
  position: absolute;
  z-index: 1000000;
  box-sizing: border-box;
  
	.s25-search-results {
		width: 100%;
		overflow-y: auto;
		position: relative;
		box-sizing: border-box;
	}
	
	.s25-options {
		width: 100%;
		
		.s25-item {
			display: block;
			width: 100%;
			box-sizing: border-box;
		}
	}
}

.s25-no-search-results {
  padding: $dropdown-padding;

}

.s25-search-results-message {
  padding: $dropdown-padding;
}

.s25-search-results-loading {
  /*! keep me */
}

.s25-search-results-minimum-error {
  /*! keep me */
}

.s25-search-results {
  /*! keep me */

}

.s25-open {
  /*! keep me */

}

.s25-search-container {
  /*! keep me */
}

/* skin */

.s25-control {
  border: $border;
  background-color: #fff;
  border-radius: $border-radius;

	&.s25-focused {
		border: $border-focus;
	
		&.s25-open {
			border-bottom: none;
			border-radius: $border-radius $border-radius 0 0;
		}
	}
}

.s25-control .s25-body {
  margin: 3px;
  margin-right: 60px;
  width: calc(100% - 60px);
}

.s25-search:focus {
  outline: 0;
}

.s25-multi .s25-toggle {
  width: 30px;
}

.s25-toggle > svg {
  display: inline-block;
  margin: auto;
}

.s25-multi {
  /*! keep me */

}

.s25-multi .s25-search {
  font-size: 100%;
  line-height: 24px;
}

.s25-multi-values {
  padding: 0;
}

.s25-content {
  /*! keep me */
}

.s25-multi-values:focus {
  outline: 0
}

.s25-multi-values .s25-item {
  position: relative;
  padding: 5px 12px 5px 18px;
  border-radius: $border-radius;
  background-color: $color-item-bkg;
  color: $color-item-text;
  margin-right: 3px;
  margin-bottom: 3px;
  transition: background .2s;
  
  	&:hover {
	  	background-color: darken( $color-item-bkg, 5% );
  	}
  
	&.s25-active {
		box-shadow: inset 0 0 0 3px #67B3FF;
	}

	&.s25-item.s25-selected {
		text-align: right;
		
		&:before {
			content: "✓";
			position: absolute;
			left: 5px;
		}
	}
}

.s25-multi-values .s25-label {
  color: rgb(34, 34, 34);
  font-size: 85%;
  border-radius: $border-radius;
  padding: 3px 3px 3px 6px;
}

.s25-multi .s25-body .s25-remove {
  border: none;
  border-radius: 0 $border-radius $border-radius 0;
  margin: 0;
  padding: 0 5px;
  outline: none;
  background-color: #f1f1f1;
  transition: background .2s;
	
	span svg {
		fill: #ccc;
		transition: fill .2s;
	}
	
	&:not(:disabled) {
		background-color: lighten( $color-item-bkg, 50% );
	
		&:hover {
		  	background-color: lighten( $color-item-bkg, 45% );
		  	
	  	}
	  	
  		span svg {
			fill: $color-item-bkg;
		}	
	}	
		
	&:focus {
		border: $border-focus;
	}
	
}

.s25-dropdown {
  border: $border-focus;
  border-top: $border;
  border-radius: 0 0 $border-radius $border-radius;
  background-color: rgb(255, 255, 255);


	.s25-body {
		padding-top: 4px;
		padding-bottom: 4px;
	}
}

.s25-live {
  border-radius: $border-radius;
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  min-height: 40px;
  max-height: 50px;
  border: 1px solid purple;
  overflow: scroll;
}

.s25-dropdown .s25-options {

	.s25-item {
		background-color: transparent;
		color: inherit;
		font-size: inherit;
		padding: $dropdown-padding;
	}

	.s25-active {
		background-color: lighten( $color-item-bkg, 50% );
	}
}

