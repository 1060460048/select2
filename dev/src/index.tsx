import * as React from "react";
import * as ReactDOM from "react-dom";

import countries from "./countries";
import * as select2 from "../../src/select2";
import {MultiSelect} from "../../src/select2";

type Country = { code: string, name: string };

const query = (function () {
    let previousTimeout: number = undefined;
    let previousReject: (reason: any) => any = undefined;

    return function (search: string, page: number) {
        if (previousReject != undefined) {
            previousReject(new Error("cancelled"));
            window.clearTimeout(previousTimeout);
        }
        const delay = 100;

        return new Promise<select2.QueryResult<Country>>((resolve: any, reject: any) => {
            previousReject = reject;
            previousTimeout = window.setTimeout(function () {
                const results: Country[] = [];
                let count = 0;
                let limit = 5;
                let offset = page * limit;
                for (let i = 0; i < countries.length; i++) {
                    const country = countries[i];
                    if (country.name.toLowerCase().indexOf(search.toLowerCase()) >= 0) {
                        if (count >= offset) {
                            results.push(country);
                        }
                        count++;
                        if (count > offset + limit) {
                            break;
                        }
                    }
                }
                previousTimeout = undefined;
                previousReject = undefined;
                resolve({values: results, more: results.length == limit})
            }, delay);
        });
    }
})();

ReactDOM.render(
    <MultiSelect
        name="countries"
        values={countries.splice(0, 3)}
        itemId="code"
        itemLabel={(c) => c.name}
        valueContent={(item)=><div style={{color:"green"}}>{item.name}</div>}
        resultContent={(item)=><div style={{color:"blue"}}>{item.name}</div>}
        valuesLabel="Countries"
        searchLabel="Add Country"
        query={query}
        minimumCharacters={2}
        openOnSearchFocus={true}
        quietMillis={100}

    />
    , document.getElementById("multi1")
);

ReactDOM.render(
    <MultiSelect
        name="countries2"
        values={countries.splice(0, 2)}
        itemId={(c) => c.code}
        itemLabel={(c) => c.name}
        valueContent={(c) => c.name}

        valuesLabel="Countries"
        searchLabel="Add Country"
        query={query}
        openOnSearchFocus={false}
    />
    , document.getElementById("multi2")
);
