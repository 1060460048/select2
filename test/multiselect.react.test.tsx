import * as React from "react";
import {MultiSelect} from "../src/select2";
import {ReactWrapper, shallow, ShallowWrapper, mount} from "enzyme";
import toJson from "enzyme-to-json";
import {makeQuery, waitForState} from "./utils";


async function search(component: ShallowWrapper | ReactWrapper, text: string) {
    component.find("input.s25Search").simulate("change", {target: {value: text}});
    await waitForState(component, (state) => {
        return state.searchResultsLoading===false;
    });
}


test("Initial Render", () => {
    let component = mount(
        <MultiSelect
            name="test"
            values={[]}
            itemId="id"
            itemLabel="text"
            valuesLabel="Values Label"
            searchLabel="Search Label"
            query={makeQuery(0)}
        />);

    expect(toJson(component)).toMatchSnapshot();

});

test("No Results", async () => {
    let component = mount(
        <MultiSelect
            name="test"
            values={[]}
            itemId="id"
            itemLabel="text"
            valuesLabel="Values Label"
            searchLabel="Search Label"
            query={makeQuery(0)}
        />);

    await search(component, "abc");

    expect(toJson(component)).toMatchSnapshot();
});

test("Some Results", async () => {
    let component = mount(
        <MultiSelect
            name="test"
            values={[]}
            itemId="id"
            itemLabel="text"
            valuesLabel="Values Label"
            searchLabel="Search Label"
            query={makeQuery(2)}
        />);

    await search(component, "abc");

    expect(toJson(component)).toMatchSnapshot();
});

