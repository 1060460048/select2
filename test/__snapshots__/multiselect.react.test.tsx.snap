// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Render 1`] = `
<MultiSelect
  dictionary="en_us"
  itemId="id"
  itemLabel="text"
  minimumCharacters={0}
  name="test"
  openOnSearchFocus={true}
  query={[Function]}
  quietMillis={-1}
  searchLabel="Search Label"
  values={Array []}
  valuesLabel="Values Label"
>
  <div
    className="s25Multi s25Control"
    id="s2-multi-0"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <div
      className="s25Body"
    >
      <button
        aria-disabled={true}
        className="s25Remove s25Offscreen"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        title="Remove selected values"
      >
        <span>
          <IconRemove
            height={20}
            width={20}
          >
            <svg
              focusable="false"
              height={20}
              tabIndex={-1}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </IconRemove>
        </span>
      </button>
      <label
        className="s25Offscreen"
        htmlFor="s2-multi-0-search"
      >
        Search Label
      </label>
      <AutosizeInput
        aria-autocomplete="none"
        aria-busy={false}
        aria-controls="s2-multi-0-dropdown"
        aria-expanded={false}
        aria-haspopup="true"
        aria-owns="s2-multi-0-dropdown"
        id="s2-multi-0-search"
        injectStyles={true}
        inputClassName="s25Search"
        inputRef={[Function]}
        inputStyle={Object {}}
        minWidth={1}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        value=""
      >
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="none"
            aria-busy={false}
            aria-controls="s2-multi-0-dropdown"
            aria-expanded={false}
            aria-haspopup="true"
            aria-owns="s2-multi-0-dropdown"
            className="s25Search"
            id="s2-multi-0-search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "2px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </AutosizeInput>
      <div
        aria-hidden={true}
        className="s25Toggle"
        onClick={[Function]}
        onMouseDownCapture={[Function]}
      >
        <IconToggle
          height={20}
          width={20}
        >
          <svg
            focusable="false"
            height={20}
            tabIndex={-1}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </IconToggle>
      </div>
    </div>
  </div>
</MultiSelect>
`;

exports[`No Results 1`] = `
<MultiSelect
  dictionary="en_us"
  itemId="id"
  itemLabel="text"
  minimumCharacters={0}
  name="test"
  openOnSearchFocus={true}
  query={[Function]}
  quietMillis={-1}
  searchLabel="Search Label"
  values={Array []}
  valuesLabel="Values Label"
>
  <div
    className="s25Multi s25Control s25SearchResultsLoading s25Open"
    id="s2-multi-1"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <div
      className="s25Body"
    >
      <button
        aria-disabled={true}
        className="s25Remove s25Offscreen"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        title="Remove selected values"
      >
        <span>
          <IconRemove
            height={20}
            width={20}
          >
            <svg
              focusable="false"
              height={20}
              tabIndex={-1}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </IconRemove>
        </span>
      </button>
      <label
        className="s25Offscreen"
        htmlFor="s2-multi-1-search"
      >
        Search Label
      </label>
      <AutosizeInput
        aria-autocomplete="none"
        aria-busy={true}
        aria-controls="s2-multi-1-dropdown"
        aria-expanded={true}
        aria-haspopup="true"
        aria-owns="s2-multi-1-dropdown"
        id="s2-multi-1-search"
        injectStyles={true}
        inputClassName="s25Search"
        inputRef={[Function]}
        inputStyle={Object {}}
        minWidth={1}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        value="abc"
      >
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="none"
            aria-busy={true}
            aria-controls="s2-multi-1-dropdown"
            aria-expanded={true}
            aria-haspopup="true"
            aria-owns="s2-multi-1-dropdown"
            className="s25Search"
            id="s2-multi-1-search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "2px",
              }
            }
            value="abc"
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            abc
          </div>
        </div>
      </AutosizeInput>
      <div
        aria-hidden={true}
        className="s25Toggle"
        onClick={[Function]}
        onMouseDownCapture={[Function]}
      >
        <IconToggle
          height={20}
          width={20}
        >
          <svg
            focusable="false"
            height={20}
            tabIndex={-1}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </IconToggle>
      </div>
    </div>
    <Dropdown
      className="s25Dropdown"
      control={
        <div
          class="s25Multi s25Control s25Open"
          id="s2-multi-1"
        >
          <div
            class="s25Body"
          >
            <button
              aria-disabled="true"
              class="s25Remove s25Offscreen"
              disabled=""
              title="Remove selected values"
            >
              <span>
                <svg
                  focusable="false"
                  height="20"
                  tabindex="-1"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </span>
            </button>
            <label
              class="s25Offscreen"
              for="s2-multi-1-search"
            >
              Search Label
            </label>
            <div
              style="display: inline-block;"
            >
              <input
                aria-autocomplete="none"
                aria-busy="false"
                aria-controls="s2-multi-1-dropdown"
                aria-expanded="true"
                aria-haspopup="true"
                aria-owns="s2-multi-1-dropdown"
                class="s25Search"
                id="s2-multi-1-search"
                role="combobox"
                style="box-sizing: content-box; width: 2px;"
                value="abc"
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              >
                abc
              </div>
            </div>
            <div
              aria-hidden="true"
              class="s25Toggle"
            >
              <svg
                focusable="false"
                height="20"
                tabindex="-1"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      }
    >
      <Portal
        containerInfo={
          <div
            class="s25Dropdown"
            style="
            top: 0px;
            left: 0px;
            width: 0px;
        "
          >
            <div
              class="s25Body"
            >
              <div
                aria-busy="false"
                class="s25SearchResults"
                style="max-height: 120px;"
              >
                <div
                  class="s25NoSearchResults"
                >
                  No results available
                </div>
              </div>
            </div>
          </div>
        }
      >
        <div
          className="s25Body"
        >
          <Results
            activeSearchResult={-1}
            dictionary={AmericanEnglishDictionary {}}
            itemLabel={[Function]}
            listboxDomId="s2-multi-1-dropdown"
            minimumCharacters={0}
            onActivateSearchResult={[Function]}
            onLoadMore={[Function]}
            onSelectSearchResult={[Function]}
            resultContent={[Function]}
            search="abc"
            searchHasMoreResults={false}
            searchResultDomId={[Function]}
            searchResultsLoading={true}
            showLoadMoreResults={false}
            showMinimumCharactersError={false}
            showNoSearchResultsFound={false}
          >
            <div
              aria-busy={true}
              className="s25SearchResults"
              onScroll={[Function]}
              style={
                Object {
                  "maxHeight": "120px",
                }
              }
            >
              <div
                className="s25SearchResultsLoading s25SearchResultsMessage"
              >
                Loading...
              </div>
            </div>
          </Results>
        </div>
      </Portal>
    </Dropdown>
  </div>
</MultiSelect>
`;

exports[`Some Results 1`] = `
<MultiSelect
  dictionary="en_us"
  itemId="id"
  itemLabel="text"
  minimumCharacters={0}
  name="test"
  openOnSearchFocus={true}
  query={[Function]}
  quietMillis={-1}
  searchLabel="Search Label"
  values={Array []}
  valuesLabel="Values Label"
>
  <div
    className="s25Multi s25Control s25SearchResultsLoading s25Open"
    id="s2-multi-2"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <div
      className="s25Body"
    >
      <button
        aria-disabled={true}
        className="s25Remove s25Offscreen"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        title="Remove selected values"
      >
        <span>
          <IconRemove
            height={20}
            width={20}
          >
            <svg
              focusable="false"
              height={20}
              tabIndex={-1}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </IconRemove>
        </span>
      </button>
      <label
        className="s25Offscreen"
        htmlFor="s2-multi-2-search"
      >
        Search Label
      </label>
      <AutosizeInput
        aria-autocomplete="none"
        aria-busy={true}
        aria-controls="s2-multi-2-dropdown"
        aria-expanded={true}
        aria-haspopup="true"
        aria-owns="s2-multi-2-dropdown"
        id="s2-multi-2-search"
        injectStyles={true}
        inputClassName="s25Search"
        inputRef={[Function]}
        inputStyle={Object {}}
        minWidth={1}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        value="abc"
      >
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="none"
            aria-busy={true}
            aria-controls="s2-multi-2-dropdown"
            aria-expanded={true}
            aria-haspopup="true"
            aria-owns="s2-multi-2-dropdown"
            className="s25Search"
            id="s2-multi-2-search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "2px",
              }
            }
            value="abc"
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            abc
          </div>
        </div>
      </AutosizeInput>
      <div
        aria-hidden={true}
        className="s25Toggle"
        onClick={[Function]}
        onMouseDownCapture={[Function]}
      >
        <IconToggle
          height={20}
          width={20}
        >
          <svg
            focusable="false"
            height={20}
            tabIndex={-1}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </IconToggle>
      </div>
    </div>
    <Dropdown
      className="s25Dropdown"
      control={
        <div
          class="s25Multi s25Control s25Open"
          id="s2-multi-2"
        >
          <div
            class="s25Body"
          >
            <button
              aria-disabled="true"
              class="s25Remove s25Offscreen"
              disabled=""
              title="Remove selected values"
            >
              <span>
                <svg
                  focusable="false"
                  height="20"
                  tabindex="-1"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </span>
            </button>
            <label
              class="s25Offscreen"
              for="s2-multi-2-search"
            >
              Search Label
            </label>
            <div
              style="display: inline-block;"
            >
              <input
                aria-activedescendant="s2-multi-2-result-3"
                aria-autocomplete="none"
                aria-busy="false"
                aria-controls="s2-multi-2-dropdown"
                aria-expanded="true"
                aria-haspopup="true"
                aria-owns="s2-multi-2-dropdown"
                class="s25Search"
                id="s2-multi-2-search"
                role="combobox"
                style="box-sizing: content-box; width: 2px;"
                value="abc"
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              >
                abc
              </div>
            </div>
            <div
              aria-hidden="true"
              class="s25Toggle"
            >
              <svg
                focusable="false"
                height="20"
                tabindex="-1"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      }
    >
      <Portal
        containerInfo={
          <div
            class="s25Dropdown"
            style="
            top: 0px;
            left: 0px;
            width: 0px;
        "
          >
            <div
              class="s25Body"
            >
              <div
                aria-busy="false"
                class="s25SearchResults"
                style="max-height: 120px;"
              >
                <div
                  aria-activedescendant="s2-multi-2-result-3"
                  class="s25Options"
                  id="s2-multi-2-dropdown"
                  role="listbox"
                >
                  <div
                    aria-label="abc-0"
                    aria-posinset="1"
                    aria-selected="true"
                    class="s25Item s25Active"
                    id="s2-multi-2-result-3"
                    role="option"
                  >
                    abc-0
                  </div>
                  <div
                    aria-label="abc-1"
                    aria-posinset="2"
                    aria-selected="false"
                    class="s25Item"
                    id="s2-multi-2-result-4"
                    role="option"
                  >
                    abc-1
                  </div>
                </div>
                <div
                  class="s25SearchResultsLoading s25SearchResultsMessage"
                >
                  Loading...
                </div>
              </div>
            </div>
          </div>
        }
      >
        <div
          className="s25Body"
        >
          <Results
            activeSearchResult={-1}
            dictionary={AmericanEnglishDictionary {}}
            itemLabel={[Function]}
            listboxDomId="s2-multi-2-dropdown"
            minimumCharacters={0}
            onActivateSearchResult={[Function]}
            onLoadMore={[Function]}
            onSelectSearchResult={[Function]}
            resultContent={[Function]}
            search="abc"
            searchHasMoreResults={false}
            searchResultDomId={[Function]}
            searchResultsLoading={true}
            showLoadMoreResults={false}
            showMinimumCharactersError={false}
            showNoSearchResultsFound={false}
          >
            <div
              aria-busy={true}
              className="s25SearchResults"
              onScroll={[Function]}
              style={
                Object {
                  "maxHeight": "120px",
                }
              }
            >
              <div
                className="s25SearchResultsLoading s25SearchResultsMessage"
              >
                Loading...
              </div>
            </div>
          </Results>
        </div>
      </Portal>
    </Dropdown>
  </div>
</MultiSelect>
`;
